monai/data/dataset.py

```diff
@@ -810,12 +811,11 @@ def _compute_cache():
             # only compute cache for the unique items of dataset
             mapping = {self.hash_func(v): v for v in data}
             self.data = list(mapping.values())
```

Comment:
So here self.data becomes of shorter length (with duplicates removed).  Then on line 817 we call _compute_cache() which starts parallel Threads for each item in self.data (via _load_cache_item func).    

But on line 818 self.data = data becomes the full list again (with duplicates).  If this assignment happens before paraller Threads are executed, wouldn't it potentially have unpredictable behaviour of caching wrong data for each hash.  A list of returned hashed data could be with duplicates, and the list of hashes (self._hash_keys) will be without.  Then using such hashes will return wrong result (potentially). 

Recommendations: you can try to deepcopy(data) and pass it as a parameter to _compute_cache() and it's sub-functions.  Also unify some sub-fuctions (_compute_cache is called only once, there is no need for it to be a subfunction).   Alternatively  we could keep existing logic (without this PR), (and for shared memory, pass it as Union[ListProxy, DictProxy] single parameter. It will be up to a user to supply the correct cashe object in case of using hash.

Reply:
is this not an issue in the most recent update?

Reply:
Hi @myron ,

The multi-thread `imap` logic here is `blocking`:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/data/dataset.py#L825
Refer to: https://docs.python.org/3/library/multiprocessing.html#multiprocessing.pool.Pool.map
So I think the assignment will not happen before parallel threads are executed.

Thanks.